traversal/procedure/ProcedureVertex.java

```diff
@@ -175,124 +189,347 @@ <ORDER extends Order> Forwardable<? extends ThingVertex, ORDER> iterateAndFilter
         ) {
-            return filterPredicates(iterator, parameters, null);
+            TreeSet<ThingVertex> filtered = new TreeSet<>();
+            vertices.forEach(v -> {
+                if (checkPredicates(v, params, set())) filtered.add(mapper.apply(v));
+            });
+            return filtered;
         }
 
-        <ORDER extends Order> Forwardable<? extends AttributeVertex<?>, ORDER> filterPredicates(
-                Forwardable<? extends AttributeVertex<?>, ORDER> iterator, Traversal.Parameters parameters,
-                @Nullable Predicate.Value<?, ?> exclude
+        <ORDER extends Order> Forwardable<KeyValue<ThingVertex, ThingVertex>, ORDER> iterateAndFilterPredicatesOnEdges(
+                List<KeyValue<ThingVertex, ThingVertex>> edges, Traversal.Parameters params, ORDER order
         ) {
-            // TODO we should be using forward() to optimise filtering for >, <, and =
-            assert id().isVariable();
-            for (Predicate.Value<?, ?> predicate : props().predicates()) {
-                if (Objects.equals(predicate, exclude)) continue;
-                for (Traversal.Parameters.Value value : parameters.getValues(id().asVariable(), predicate)) {
-                    iterator = iterator.filter(a -> predicate.apply(a.asAttribute(), value));
+            if (props().predicates().isEmpty()) return iterateSorted(edges, order);
+            else {
+                if (iterate(edges).anyMatch(kv -> kv.key().isAttribute() && kv.key().asAttribute().type().valueType().equals(STRING))) {
+                    return iterateSorted(filterPredicatesAndMapEdges(edges, params, ThingVertex::asAttribute), order);
+                } else {
+                    return iterateSorted(filterPredicatesAndMapEdges(edges, params, v -> v.asAttribute().toValue()), order);
                 }
             }
-            return iterator;
         }
 
-        Forwardable<KeyValue<ThingVertex, ThingVertex>, Order.Asc> filterPredicatesOnEdge(
-                Forwardable<KeyValue<ThingVertex, ThingVertex>, Order.Asc> iterator, Traversal.Parameters parameters
+        private TreeSet<KeyValue<ThingVertex, ThingVertex>> filterPredicatesAndMapEdges(
+                List<KeyValue<ThingVertex, ThingVertex>> edges,
+                Traversal.Parameters params, Function<ThingVertex, AttributeVertex<?>> mapper
+        ) {
+            TreeSet<KeyValue<ThingVertex, ThingVertex>> filtered = new TreeSet<>();
+            edges.forEach(kv -> {
+                if (checkPredicates(kv.key(), params, set())) {
+                    filtered.add(new KeyValue<>(mapper.apply(kv.key()), kv.value()));
+                }
+            });
+            return filtered;
+        }
+
+        <ORDER extends Order> Forwardable<? extends ThingVertex, ORDER> mergeAndFilterPredicatesOnVertices(
+                GraphManager graphMgr, List<Pair<TypeVertex, Forwardable<ThingVertex, ORDER>>> vertexIters,
```

Comment:
a list of `Forwardable` iterators to be merged together

Reply:
each `Forwardable` iterator comes in a pair, where the type vertex is provided with the iterator to denote the type of the vertices that will be returned by that iterator